framework/src/st_things/things_stack/src/common/utils/things_ping.c

```diff
@@ -0,0 +1,1255 @@
+/****************************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#define OICPING_ENABLE      1
+
+#include <stdbool.h>
+#include <unistd.h>
+#include <pthread.h>
+#include "ocpayload.h"
+#include "octypes.h"
+#include "es_common.h"
+#include "things_malloc.h"
+#include "things_logger.h"
+#include "things_list.h"
+#include "things_ping.h"
+
+#ifdef __ST_THINGS_RTOS__
+#include "things_rtos_util.h"
+#endif
+
+#ifdef OICPING_ENABLE
+#include "oickeepalive.h"
+#endif
+#include "resource_handler.h"
+#include "cloud_manager.h"
+
+#include "things_thread.h"
+
+#define TAG "[things_ping]"
+
+#define DEFAULT_CONTEXT_VALUE   0x99
+#define PING_UNIT           60	// 60 seconds is minimum value of ping.
+#define NET_TROUBLE_DELAY   5	// 5 seconds
+#define MAX_INTERVAL_EA     4	// Cloud Support 4 EA Interval class.
+
+#define KEEPALIVE_RESOURCE_URI "/oic/ping"
+
+typedef void *(*pthread_func_type)(void *);
+typedef struct things_ping_s {
+	pthread_mutex_t mutex_state;	// mutex for State variable.
+	ping_state_e bit_mask_state;	// State of Ping
+
+	pthread_mutex_t mutex_thread;	// mutex for handle_thread variable.
+	pthread_t handle_thread;	// Ping-Thread Handler
+	bool continue_thread;		// Continue Ping-Thread Flag.
+	char *addr;					// IP address of remote
+	uint16_t port;				// Port number for remote
+
+	pthread_mutex_t mutex_int;	// mutex for interval variable.
+	int64_t interval[MAX_INTERVAL_EA];	// Interval EA  (unit : minute , max : 60)
+	int int_seq;				// sequence of interval
+} things_ping_s;
+
+static list_s *list = NULL;
+static const char INTERVAL_ARRAY[] = "inarray";
+static const char INTERVAL[] = "in";
+things_ping_s *pPing = NULL;	//SKKIM for ping command
```

Comment:
If pPing is not used, please remove this variable.

Reply:
modified